teps/0130-pipeline-level-service.md

```diff
@@ -0,0 +1,759 @@
+---
+status: proposed
+title: Pipeline-level Service
+creation-date: '2023-01-24'
+last-updated: '2023-02-21'
+authors:
+- '@lbernick'
+collaborators: []
+---
+
+# TEP-0130: Pipeline-level Service
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Use Cases](#use-cases)
+    - [Docker builds](#docker-builds)
+    - [Other uses for this feature](#other-uses-for-this-feature)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Requirements](#requirements)
+  - [Related features in other CI/CD systems](#related-features-in-other-cicd-systems)
+- [User Experience](#user-experience)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [Conformance](#conformance)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Improve our documentation](#improve-our-documentation)
+  - [Catalog Tasks for docker daemon](#catalog-tasks-for-docker-daemon)
+  - [Pipeline-level Sidecar](#pipeline-level-sidecar)
+    - [Open questions](#open-questions)
+    - [Notes: Reusable sidecars](#notes-reusable-sidecars)
+  - [First-class docker support](#first-class-docker-support)
+  - [Docker daemon Custom Task](#docker-daemon-custom-task)
+  - [Allow Pipeline Tasks to be run as daemons](#allow-pipeline-tasks-to-be-run-as-daemons)
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes support for managing services with the same lifespan as a PipelineRun.
+The primary use case for this feature is a Pipeline-level Docker daemon,
+since it will allow users to create more customized Docker build Pipelines.
+
+## Motivation
+
+It's challenging to write a Tekton Pipeline that depends on an external service without exposing the service to multiple PipelineRuns. 
+Task sidecars meet needs such as integration testing, but sometimes a service needs to be shared between multiple Tasks in a PipelineRun.
+Right now, the best way for users to create a Pipeline-level service is to have one Task that spins up the service and a Finally Task that
+tears it down, but this is non-trivial to write and leads to poor separation of concerns (app developers writing build Pipelines must also handle
+infrastructure details of spinning up the service).
```

Comment:
I rephrased this a bit, but I think the main point I'm trying to get across is that you have to build the infrastructure directly into your pipeline, when different people might be responsible for these components. You're right that it depends on architecture whether the app developer interacts with the k8s cluster or not; what I'm trying to say is more that we should make it easy to separate this out for companies where it would be two different roles. Is there a better way of phrasing this you have in mind?